rfc/legacy_artifacts_ui_removal.md

```diff
@@ -0,0 +1,131 @@
+
+### Goals
+The goal of this RFC is to propose a plan to:
+- Test the new UI.
+- Resolve all bugs reported by users or discovered while testing the new UI.
+- Enable the new UI by default.
+- Communicate the new default to users.
+- Remove the old UI.
+
+### Non-Goals
+
+- Compare the implementation details of the legacy and new UIs.
+- Gather user feedback about the new UI in order to further improve it: while
+this might be worthwhile, the scope of this RFC is limited to auditing the new
+UI as-is to ensure that it is backwards-compatible with the legacy UI.
+
+## Motivation and Rationale
+
+- Pivotal wrote this
+[design document](https://docs.google.com/document/d/10y1KIBoSWlerFxkkW8nJLZ_zspz7ToCRifnWoFHyp8k/edit?usp=sharing)
+for improving the artifact configuration experience, and released an initial
+implementation behind a feature flag in Spinnaker 1.13 (March 2019). However,
+the design document did not include a plan for identifying validating users,
+exposing the new UI by default, or removing support for the legacy UI.
+- Pivotal originally
+[intended to default the new UI to enabled](https://github.com/spinnaker/spinnaker/issues/4777)
+in Spinnaker 1.16, but Pivotal is no longer working on Spinnaker.
+- Supporting two artifacts-related feature flags has increased the complexity of
+the artifacts configuration experience, which was already confusing for end
+users, as evidenced by many conversations in Slack, GitHub issues, and the
+Kubernetes SIG. There have also been a number of bugs reported in relation to
+the new UI, but no comprehensive audit of the new UI by Spinnaker maintainers.
+
+## Timeline
+
+Each set of milestones should be complete by the time the corresponding Spinnaker version is released.
+
+| Spinnaker Version | Release Window Opens | Milestones |
+|-------------------|----------------------|------------|
+| 1.20              | 2020-04-27           | New UI is enabled by default. Legacy UI is behind a feature flag.
+| 1.21              | 2020-06-22           | Legacy UI is removed.
+| 1.23              | TBD                  | Legacy UI documentation is removed.
+
+## Design
+
+- [x] Create a
+[GitHub project](https://github.com/spinnaker/spinnaker/projects/17)
+to organize existing user-reported issues related to the new UI.
+
+- [] Manually test the new UI. For each type of artifact and artifact-driven
+workflow, test that pipelines configured with the legacy UI still work with the
+new UI, and that pipelines configured with the new UI work as expected. This
+includes artifacts configured in the pipeline configuration's Expected Artifacts
+section, artifacts configured as expected artifacts inline in stages, and
+artifacts configured in Produces Artifacts sections of stages.
+
+- [] Document and resolve all bugs discovered while manually testing the new UI.
+
+- [] For release 1.20, remove Deck's `artifacts` and `artifactsRewrite` feature
+flags from both Deck and Halyard. The default UI will be the new UI,
```

Comment:
What happens when you remove them? Will there be an error? It would probably be nice if there was something that said "you were using the `artifacts` feature, but that's deprecated now, the new thing is available instead, or you can use `legacyArtifactsEnabled` to turn the old crappy thing back on (but must migrate before Spinnaker 1.21)!"

I guess maybe that happens automatically via the `tooHighMessage` parameter to `@ValidForSpinnakerVersion`? If so, that's great, maybe mention that will happen under "Community this change by..."?

Reply:
Thanks for calling this out, @plumpy! I updated the RFC to clarify that we will annotate the existing artifacts feature flags with `@ValidForSpinnakerVersion` with an appropriate `tooHighMessage` for release 1.20, and only remove them after release 1.23.